pkg/cmd/admin/migrate/images/imagerefs.go

```diff
@@ -0,0 +1,475 @@
+* buildconfigs         * replicationcontrollers
+* deploymentconfigs    * pods
+* secrets (docker)
+
+Only images, imagestreams, and secrets are updated by default. Updating images and image
+streams requires administrative privileges.`
+
+	internalMigrateImagesExample = `  # Perform a dry-run of migrating all "docker.io" references to "myregistry.com"
+  %[1]s docker.io/*=myregistry.com/*
+
+  # To actually perform the migration, the confirm flag must be appended
+  %[1]s docker.io/*=myregistry.com/* --confirm
+
+  # To see more details of what will be migrated, use the loglevel and output flags
+  %[1]s docker.io/*=myregistry.com/* --loglevel=2 -o yaml
+
+  # Migrate from a service IP to an internal service DNS name
+  %[1]s 172.30.1.54/*=registry.openshift.svc.cluster.local/*
+
+  # Migrate from a service IP to an internal service DNS name for all deployment configs and builds
+  %[1]s 172.30.1.54/*=registry.openshift.svc.cluster.local/* --include=buildconfigs,deploymentconfigs`
+)
+
+type MigrateImageReferenceOptions struct {
+	migrate.ResourceOptions
+
+	Client          client.Interface
+	Mappings        ImageReferenceMappings
+	UpdatePodSpecFn func(obj runtime.Object, fn func(*kapi.PodSpec) error) (bool, error)
+}
+
+// NewCmdMigrateImageReferences implements a MigrateImages command
+func NewCmdMigrateImageReferences(name, fullName string, f *clientcmd.Factory, in io.Reader, out, errout io.Writer) *cobra.Command {
+	options := &MigrateImageReferenceOptions{
+		ResourceOptions: migrate.ResourceOptions{
+			In:      in,
+			Out:     out,
+			ErrOut:  errout,
+			Include: []string{"imagestream", "image", "secrets"},
+		},
+	}
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s REGISTRY/NAME=REGISTRY/NAME [...]", name),
+		Short:   "Update embedded Docker image references",
+		Long:    internalMigrateImagesLong,
+		Example: fmt.Sprintf(internalMigrateImagesExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(options.Complete(f, cmd, args))
+			kcmdutil.CheckErr(options.Validate())
+			kcmdutil.CheckErr(options.Run())
+		},
+	}
+	options.ResourceOptions.Bind(cmd)
+
+	return cmd
+}
+
+func (o *MigrateImageReferenceOptions) Complete(f *clientcmd.Factory, c *cobra.Command, args []string) error {
+	var remainingArgs []string
+	for _, s := range args {
+		if !strings.Contains(s, "=") {
+			remainingArgs = append(remainingArgs, s)
+			continue
+		}
+		mapping, err := ParseMapping(s)
+		if err != nil {
+			return err
+		}
+		o.Mappings = append(o.Mappings, mapping)
+	}
+
+	o.UpdatePodSpecFn = f.UpdatePodSpecForObject
+
+	if len(remainingArgs) > 0 {
+		return fmt.Errorf("all arguments must be valid FROM=TO mappings")
+	}
+
+	o.ResourceOptions.SaveFn = o.save
+	if err := o.ResourceOptions.Complete(f, c); err != nil {
+		return err
+	}
+
+	osclient, _, err := f.Clients()
+	if err != nil {
+		return err
+	}
+	o.Client = osclient
+
+	return nil
+}
+
+func (o *MigrateImageReferenceOptions) Validate() error {
```

Comment:
That's strange, because it's not sufficient to defend against any problem
(mutation of sub resources or interfaces), so it's kind of arbitrary, and
goes against our general conventions of always passing pointers unless we
have a real reason not to.

On Wed, Jul 27, 2016 at 4:01 PM, Maciej Szulik <notifications@github.com>
wrote:

> In pkg/cmd/admin/migrate/images/imagerefs.go
> <https://github.com/openshift/origin/pull/9240#discussion_r72512049>:
>
> > +
> > +	o.ResourceOptions.SaveFn = o.save
> > +	if err := o.ResourceOptions.Complete(f, c); err != nil {
> > +		return err
> > +	}
> > +
> > +	osclient, _, err := f.Clients()
> > +	if err != nil {
> > +		return err
> > +	}
> > +	o.Client = osclient
> > +
> > +	return nil
> > +}
> > +
> > +func (o *MigrateImageReferenceOptions) Validate() error {
>
> According to our conventions
> <https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md#command-implementation-conventions>
> both Validate and Run should non-pointer methods.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9240/files/1ddd7d050a21f14d07b8d4198df6e7eec9cf8475#r72512049>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p2L_GGWc9FhgS5T9vXhxerrvLGOSks5qZ7ktgaJpZM4IxnZA>
> .
>